### ISSUE TITLE
KBestExtractor.getKthHyp(node,k) is called twice for topN=1

### ISSUE THREAD

**Issue by fhieber** [joshua-decoder/joshua#186]
Hi,

not sure if this is a bug or if I am not aware of some parts of the algorithm that handle internal state, but the for-loop in line 223 (https://github.com/joshua-decoder/joshua/blob/master/src/joshua/decoder/hypergraph/KBestExtractor.java#L223) seems incorrect to me: if `topN` is set to 1, `getKthHyp()` is executed twice. The result of the second execution is discarded due to breaking out of the loop directly afterwards.

A related question: If I am only interested in the 1-best hypothesis, do I need to set `topN` to 1 or 0, i.e. does `ViterbiExtractor` return the very same hypothesis as `getKthHyp(node, 1)`, or is there some rescoring with stateful feature functions such as language models involved during kbest extraction? It seems to me that cube pruning and everything is done beforehand (`decoder.chart_parser.chart`), correct?
Thanks for your help!

**Comment by mjpost** [comment]
Yes, it looks like the logic here should be changed.

As for the related question: Joshua does not store feature values during decoding, but only dot products. This makes things more efficient, because each edge in the hypergraph only has to store a scalar instead of a large sparse `FeatureVector` object. After the hypergraph is built, the actual feature values can be recovered, if desired, by replaying the features during k-best extraction. Firing up this code is very expensive (k-best extraction is often as time-consuming as the decoding itself, both at low pop-limit settings (where the overhead of extraction dominates) and at high pop-limit settings (where recomputing scores in a denser forest dominates)), and furthermore is only needed for tuning. So Joshua provides the option of setting `topN==0` to turn off this code entirely. When that happens, only things like the inside model score (which can be read off the top node) or the Viterbi string (which can be obtained via `ViterbiExtractor`) are available. 

In summary, if you don't care about the feature values and only need the top-level string, set `topN` to 0. In that situation, `output-format` will be ignored (though perhaps that should change, with certain values, like `%f`, just being ignored).

I hope that is clarifying!

**Comment by fhieber** [comment]
This is great, thanks for the change! 
Just for the sake of completeness: output format '%S', i.e. a de-normalized hypothesis, would also be applicable to the Viterbi hypothesis, right?

**Comment by mjpost** [comment]
Yes, it would, though it wasn't added (on account of laziness) until the commit I just pushed.
